// dmdb/src/net/buffer.js
"use strict";

var __awaiter = this && this.__awaiter || function(t, e, s, h) {
    return new (s = s || Promise)(function(n, i) {
        function r(t) {
            try {
                f(h.next(t));
            } catch (t) {
                i(t);
            }
        }
        function u(t) {
            try {
                f(h.throw(t));
            } catch (t) {
                i(t);
            }
        }
        function f(t) {
            var i;
            t.done ? n(t.value) : ((i = t.value) instanceof s ? i : new s(function(t) {
                t(i);
            })).then(r, u);
        }
        f((h = h.apply(t, e || [])).next());
    });
}, __generator = this && this.__generator || function(n, r) {
    var u, f, e, s = {
        label: 0,
        sent: function() {
            if (1 & e[0]) throw e[1];
            return e[1];
        },
        trys: [],
        ops: []
    }, t = {
        next: i(0),
        throw: i(1),
        return: i(2)
    };
    return "function" == typeof Symbol && (t[Symbol.iterator] = function() {
        return this;
    }), t;
    function i(i) {
        return function(t) {
            return function(i) {
                if (u) throw new TypeError("Generator is already executing.");
                for (;s; ) try {
                    if (u = 1, f && (e = 2 & i[0] ? f.return : i[0] ? f.throw || ((e = f.return) && e.call(f), 
                    0) : f.next) && !(e = e.call(f, i[1])).done) return e;
                    switch (f = 0, e && (i = [ 2 & i[0], e.value ]), i[0]) {
                      case 0:
                      case 1:
                        e = i;
                        break;

                      case 4:
                        return s.label++, {
                            value: i[1],
                            done: !1
                        };

                      case 5:
                        s.label++, f = i[1], i = [ 0 ];
                        continue;

                      case 7:
                        i = s.ops.pop(), s.trys.pop();
                        continue;

                      default:
                        if (!(e = 0 < (e = s.trys).length && e[e.length - 1]) && (6 === i[0] || 2 === i[0])) {
                            s = 0;
                            continue;
                        }
                        if (3 === i[0] && (!e || i[1] > e[0] && i[1] < e[3])) {
                            s.label = i[1];
                            break;
                        }
                        if (6 === i[0] && s.label < e[1]) {
                            s.label = e[1], e = i;
                            break;
                        }
                        if (e && s.label < e[2]) {
                            s.label = e[2], s.ops.push(i);
                            break;
                        }
                        e[2] && s.ops.pop(), s.trys.pop();
                        continue;
                    }
                    i = r.call(n, s);
                } catch (t) {
                    i = [ 6, t ], f = 0;
                } finally {
                    u = e = 0;
                }
                if (5 & i[0]) throw i[1];
                return {
                    value: i[0] ? i[1] : void 0,
                    done: !0
                };
            }([ i, t ]);
        };
    }
};

Object.defineProperty(exports, "__esModule", {
    value: !0
});

var bufferUtil_1 = require("../utils/bufferUtil"), msg_1 = require("./msg/msg"), DmBuffer = function() {
    function t(t) {
        this._readOff = 0, this._writeOff = 0, "number" == typeof t ? (t < 0 && (t = 0), 
        this._buf = Buffer.allocUnsafe(t)) : (this._buf = Buffer.from(t), this._writeOff = t.length);
    }
    return Object.defineProperty(t.prototype, "ti", {
        get: function() {
            return this._buf.slice(0, this._writeOff);
        },
        enumerable: !0,
        configurable: !0
    }), Object.defineProperty(t.prototype, "ii", {
        get: function() {
            return this._readOff;
        },
        enumerable: !0,
        configurable: !0
    }), Object.defineProperty(t.prototype, "Gt", {
        get: function() {
            return this._writeOff;
        },
        enumerable: !0,
        configurable: !0
    }), t.prototype.Tt = function(t) {
        return this._writeOff = t, this._readOff > t && (this._readOff = t), this;
    }, t.prototype.Jt = function(t) {
        return this._readOff = t, this;
    }, t.prototype.ni = function(t) {
        return t ? this._buf.length - this._writeOff : this.Gt - this._readOff;
    }, t.prototype.ri = function(t, i, n) {
        return i ? n ? this.ui(t) : this.Tt(t) : n ? this._readOff += t : this._readOff -= t, 
        this;
    }, t.prototype.fi = function(f, e) {
        return __awaiter(this, void 0, void 0, function() {
            var u = this;
            return __generator(this, function(t) {
                return [ 2, new Promise(function(t, i) {
                    setTimeout(function() {
                        i(new Error("Socket read timeout"));
                    }, 3e3);
                    var n = null, r = setInterval(function() {
                        null !== n && (clearInterval(r), u.It(n), t(n.length));
                    }, 1);
                    f.once("readable", function() {
                        var t = f.read(e), i = t.readInt32LE(msg_1.Msg.ei);
                        0 < i && (t = Buffer.concat([ t, f.read(i) ])), n = t;
                    });
                }) ];
            });
        });
    }, t.prototype.si = function(u) {
        return __awaiter(this, void 0, void 0, function() {
            var r = this;
            return __generator(this, function(t) {
                return [ 2, new Promise(function(i, n) {
                    u.write(r._buf, function(t) {
                        t ? n(t) : (console.log("send msg"), i(r));
                    });
                }) ];
            });
        });
    }, t.prototype.hi = function() {
        return 1 <= this.ni(!1) ? (this._readOff += 1, this._buf.readInt8(this._readOff - 1)) : bufferUtil_1.BufferUtil.oi(this.ci(1));
    }, t.prototype.li = function() {
        return 1 <= this.ni(!1) ? (this._readOff += 1, this._buf.readUInt8(this._readOff - 1)) : bufferUtil_1.BufferUtil._i(this.ci(1));
    }, t.prototype.bi = function() {
        return 2 <= this.ni(!1) ? (this._readOff += 2, this._buf.readInt16LE(this._readOff - 2)) : bufferUtil_1.BufferUtil.ai(this.ci(2));
    }, t.prototype.T = function() {
        return 2 <= this.ni(!1) ? (this._readOff += 2, this._buf.readUInt16LE(this._readOff - 2)) : bufferUtil_1.BufferUtil.Dt(this.ci(2));
    }, t.prototype.D = function() {
        return 4 <= this.ni(!1) ? (this._readOff += 4, this._buf.readInt32LE(this._readOff - 4)) : bufferUtil_1.BufferUtil.Ui(this.ci(4));
    }, t.prototype.vi = function() {
        return 4 <= this.ni(!1) ? (this._readOff += 4, this._buf.readUInt32LE(this._readOff - 4)) : bufferUtil_1.BufferUtil.mi(this.ci(4));
    }, t.prototype.gi = function() {
        return 8 <= this.ni(!1) ? (this._readOff += 8, bufferUtil_1.BufferUtil.di(this._buf, this._readOff - 8)) : bufferUtil_1.BufferUtil.V(this.ci(8));
    }, t.prototype.wi = function() {
        return 8 <= this.ni(!1) ? (this._readOff += 8, bufferUtil_1.BufferUtil.yi(this._buf, this._readOff - 8)) : bufferUtil_1.BufferUtil.Bi(this.ci(8));
    }, t.prototype.pi = function() {
        return 4 <= this.ni(!1) ? (this._readOff += 4, this._buf.readFloatLE(this._readOff - 4)) : bufferUtil_1.BufferUtil.ki(this.ci(4));
    }, t.prototype.Oi = function() {
        return 8 <= this.ni(!1) ? (this._readOff += 8, this._buf.readDoubleLE(this._readOff - 8)) : bufferUtil_1.BufferUtil.ji(this.ci(8));
    }, t.prototype.ci = function(t) {
        var i = bufferUtil_1.BufferUtil.m(this._buf, this._readOff, t);
        return this._readOff += t, i;
    }, t.prototype.xi = function() {
        return this.ci(this.vi());
    }, t.prototype.Pi = function() {
        return this.ci(this.li());
    }, t.prototype.Si = function() {
        return this.ci(this.T());
    }, t.prototype.qi = function() {
        for (var t = 0; 0 != this.li(); ) t++;
        return this.ri(t, !1, !1), this.ci(t);
    }, t.prototype.Di = function(t, i) {
        return bufferUtil_1.BufferUtil.G(this.ci(t), i);
    }, t.prototype.Ei = function(t) {
        return bufferUtil_1.BufferUtil.G(this.xi(), t);
    }, t.prototype.Ii = function(t) {
        return bufferUtil_1.BufferUtil.G(this.Pi(), t);
    }, t.prototype.S = function(t) {
        return bufferUtil_1.BufferUtil.G(this.Si(), t);
    }, t.prototype.Mi = function(t) {
        return bufferUtil_1.BufferUtil.G(this.qi(), t);
    }, t.prototype.Ti = function(t) {
        return this.Gi(t ? 1 : 0);
    }, t.prototype.zi = function(t) {
        return 1 <= this.ni(!0) ? (this._buf.writeInt8(t, this._writeOff), this._writeOff += 1, 
        1) : this.It(bufferUtil_1.BufferUtil.y(t));
    }, t.prototype.Gi = function(t) {
        return 1 <= this.ni(!0) ? (this._buf.writeUInt8(255 & t, this._writeOff), this._writeOff += 1, 
        1) : this.It(bufferUtil_1.BufferUtil.Ai(t));
    }, t.prototype.x = function(t) {
        return 2 <= this.ni(!0) ? (this._buf.writeInt16LE(t, this._writeOff), this._writeOff += 2, 
        2) : this.It(bufferUtil_1.BufferUtil.p(t));
    }, t.prototype.Ci = function(t) {
        return 2 <= this.ni(!0) ? (this._buf.writeUInt16LE(65535 & t, this._writeOff), this._writeOff += 2, 
        2) : this.It(bufferUtil_1.BufferUtil.Fi(t));
    }, t.prototype.k = function(t) {
        return 4 <= this.ni(!0) ? (this._buf.writeInt32LE(t, this._writeOff), this._writeOff += 4, 
        4) : this.It(bufferUtil_1.BufferUtil.h(t));
    }, t.prototype.Hi = function(t) {
        return 4 <= this.ni(!0) ? (this._buf.writeUInt32LE((4294967295 & t) >>> 0, this._writeOff), 
        this._writeOff += 4, 4) : this.It(bufferUtil_1.BufferUtil.Ji(t));
    }, t.prototype.Ki = function(t) {
        return 8 <= this.ni(!0) ? (bufferUtil_1.BufferUtil.Li(this._buf, t, this._writeOff), 
        this._writeOff += 8, 8) : this.It(bufferUtil_1.BufferUtil.l(t));
    }, t.prototype.Ni = function(t) {
        return 8 <= this.ni(!0) ? (bufferUtil_1.BufferUtil.Qi(this._buf, t, this._writeOff), 
        this._writeOff += 8, 8) : this.It(bufferUtil_1.BufferUtil.Ri(t));
    }, t.prototype.Vi = function(t) {
        return 4 <= this.ni(!0) ? (this._buf.writeFloatLE(t, this._writeOff), this._writeOff += 4, 
        4) : this.It(bufferUtil_1.BufferUtil.w(t));
    }, t.prototype.Wi = function(t) {
        return 8 <= this.ni(!0) ? (this._buf.writeDoubleLE(t, this._writeOff), this._writeOff += 8, 
        8) : this.It(bufferUtil_1.BufferUtil.v(t));
    }, t.prototype.It = function(t) {
        var i = t.length;
        if (0 == i) return i;
        var n, r, u = this._buf.length;
        return this._writeOff + i > u && (n = Math.max(2 * u, this._writeOff + i), r = Buffer.allocUnsafe(n), 
        this._buf = r.fill(this._buf, 0, this._buf.length)), this._buf = this._buf.fill(t, this._writeOff, this._writeOff + i), 
        this._writeOff += i, i;
    }, t.prototype.Xi = function(t) {
        return this.Hi(t.length) + this.It(t);
    }, t.prototype.Yi = function(t) {
        return this.Gi(t.length) + this.It(t);
    }, t.prototype.Zi = function(t) {
        return this.Ci(t.length) + this.It(t);
    }, t.prototype.$i = function(t) {
        return this.It(t) + this.Gi(0);
    }, t.prototype.tn = function(t, i) {
        var n = bufferUtil_1.BufferUtil.o(t, i);
        return this.Xi(n);
    }, t.prototype.in = function(t, i) {
        var n = bufferUtil_1.BufferUtil.o(t, i);
        return this.Yi(n);
    }, t.prototype.N = function(t, i) {
        var n = bufferUtil_1.BufferUtil.o(t, i);
        return this.Zi(n);
    }, t.prototype.nn = function(t, i) {
        var n = bufferUtil_1.BufferUtil.o(t, i);
        return this.$i(n);
    }, t.prototype.t = function(t, i) {
        return this._(t, bufferUtil_1.BufferUtil.y(i));
    }, t.prototype.rn = function(t, i) {
        return this._(t, bufferUtil_1.BufferUtil.Ai(i));
    }, t.prototype.B = function(t, i) {
        return this._(t, bufferUtil_1.BufferUtil.p(i));
    }, t.prototype.e = function(t, i) {
        return this._(t, bufferUtil_1.BufferUtil.Fi(i));
    }, t.prototype.f = function(t, i) {
        return this._(t, bufferUtil_1.BufferUtil.h(i));
    }, t.prototype.n = function(t, i) {
        return this._(t, bufferUtil_1.BufferUtil.Ji(i));
    }, t.prototype.r = function(t, i) {
        return this._(t, bufferUtil_1.BufferUtil.l(i));
    }, t.prototype.un = function(t, i) {
        return this._(t, bufferUtil_1.BufferUtil.Ri(i));
    }, t.prototype._ = function(t, i) {
        return this._buf.set(i, t), i.length;
    }, t.prototype.fn = function(t, i) {
        return this.n(t, i.length) + this._(t + 4, i);
    }, t.prototype.en = function(t, i) {
        return this.rn(t, i.length) + this._(t + 1, i);
    }, t.prototype.sn = function(t, i) {
        return this.e(t, i.length) + this._(t + 2, i);
    }, t.prototype.hn = function(t, i) {
        return this._(t, i) + this.rn(t + i.length, 0);
    }, t.prototype.on = function(t, i, n) {
        return this.fn(t, bufferUtil_1.BufferUtil.o(i, n));
    }, t.prototype.cn = function(t, i, n) {
        return this.en(t, bufferUtil_1.BufferUtil.o(i, n));
    }, t.prototype.ln = function(t, i, n) {
        return this.sn(t, bufferUtil_1.BufferUtil.o(i, n));
    }, t.prototype._n = function(t, i, n) {
        return this.hn(t, bufferUtil_1.BufferUtil.o(i, n));
    }, t.prototype.a = function(t) {
        return bufferUtil_1.BufferUtil.oi(this.m(t, 1));
    }, t.prototype.H = function(t) {
        return bufferUtil_1.BufferUtil._i(this.m(t, 1));
    }, t.prototype.c = function(t) {
        return bufferUtil_1.BufferUtil.ai(this.m(t, 2));
    }, t.prototype.U = function(t) {
        return bufferUtil_1.BufferUtil.Dt(this.m(t, 2));
    }, t.prototype.s = function(t) {
        return bufferUtil_1.BufferUtil.Ui(this.m(t, 4));
    }, t.prototype.g = function(t) {
        return bufferUtil_1.BufferUtil.mi(this.m(t, 4));
    }, t.prototype.i = function(t) {
        return bufferUtil_1.BufferUtil.V(this.m(t, 8));
    }, t.prototype.bn = function(t) {
        return bufferUtil_1.BufferUtil.Bi(this.m(t, 8));
    }, t.prototype.d = function(t) {
        return bufferUtil_1.BufferUtil.ki(this.m(t, 4));
    }, t.prototype.b = function(t) {
        return bufferUtil_1.BufferUtil.ji(this.m(t, 8));
    }, t.prototype.m = function(t, i) {
        return this._buf.subarray(t, t + i);
    }, t.prototype.an = function(t) {
        var i = this.g(t);
        return this.m(t + 4, i);
    }, t.prototype.Un = function(t) {
        var i = this.H(t);
        return this.m(t + 1, i);
    }, t.prototype.vn = function(t) {
        var i = this.U(t);
        return this.m(t + 2, i);
    }, t.prototype.mn = function(t) {
        for (var i = 0; 0 !== this.H(t); ) t++, i++;
        return this.m(t - i, i);
    }, t.prototype.gn = function(t, i) {
        return bufferUtil_1.BufferUtil.G(this.an(t), i);
    }, t.prototype.dn = function(t, i) {
        return bufferUtil_1.BufferUtil.G(this.Un(t), i);
    }, t.prototype.wn = function(t, i) {
        return bufferUtil_1.BufferUtil.G(this.vn(t), i);
    }, t.prototype.yn = function(t, i) {
        return bufferUtil_1.BufferUtil.G(this.mn(t), i);
    }, t.prototype.ui = function(t) {
        var i, n = this._buf.length;
        return this._writeOff + t > n && (i = Buffer.allocUnsafe(2 * n), this._buf = i.fill(this._buf, 0, n)), 
        this._buf.fill(0, this._writeOff, this._writeOff + t), this._writeOff += t, this;
    }, t;
}();

exports.DmBuffer = DmBuffer;
