// dmdb/src/net/msg/login.js
"use strict";

var __extends = this && this.__extends || function() {
    var h = function(t, i) {
        return (h = Object.setPrototypeOf || {
            __proto__: []
        } instanceof Array && function(t, i) {
            t.__proto__ = i;
        } || function(t, i) {
            for (var s in i) i.hasOwnProperty(s) && (t[s] = i[s]);
        })(t, i);
    };
    return function(t, i) {
        function s() {
            this.constructor = t;
        }
        h(t, i), t.prototype = null === i ? Object.create(i) : (s.prototype = i.prototype, 
        new s());
    };
}();

Object.defineProperty(exports, "__esModule", {
    value: !0
});

var msg_1 = require("./msg"), const_1 = require("../../desc/const"), connection_1 = require("../../driver/connection"), bufferUtil_1 = require("../../utils/bufferUtil"), error_1 = require("../../driver/error"), rsaCipher_1 = require("../../security/rsaCipher"), driverUtil_1 = require("../../utils/driverUtil"), driver_1 = require("../../driver"), Login = function(h) {
    function _(t, i) {
        var s = h.call(this, t, msg_1.Msg.Ms, null, i) || this;
        return s.Us = !1, s.pt = t.pt, s;
    }
    return __extends(_, h), _.processServerVersion = function(t, i) {
        if (driverUtil_1.DriverUtil.compareDBVersion(t, driver_1.Driver.VERSION_LOWEST_DATABASE) < 0) throw error_1.DBError.ECJS_ERR_SERVER_VERSION();
        i.malini2 = 0 < driverUtil_1.DriverUtil.compareDBVersion(t, driver_1.Driver.VERSION_MALINI2), 
        i.execute2 = 0 < driverUtil_1.DriverUtil.compareDBVersion(t, driver_1.Driver.VERSION_EXECUTE2), 
        i.lobEmptyCompOrcl = 0 < driverUtil_1.DriverUtil.compareDBVersion(t, driver_1.Driver.VERSION_LOB_EMPTY_COMP_ORCL), 
        i.lobOptimized = 0 < driverUtil_1.DriverUtil.compareDBVersion(t, driver_1.Driver.VERSION_SERVER_LOB_OPTIMIZED);
    }, _.prototype.jn = function() {
        var t = this.pt.getServerEncoding();
        this.St.f(_.Hs, this.pt.newClientType ? const_1.Const.SESS3_CLT_NODE : const_1.Const.SESS3_CLT_JDBC), 
        this.St.f(_.qs, connection_1.Connection.node2dbIsoLevel(this.pt.isoLevel)), this.St.f(_.Fs, this.pt.conn_prop_locale), 
        this.St.B(_.Ss, this.pt.conn_prop_localTimezone), this.St.t(_.Bs, this.pt.readOnly ? const_1.Const.TRX_ACCESS_MODE_READ_ONLY : const_1.Const.TRX_ACCESS_MODE_READ_WRITE), 
        this.St.f(_.ws, this.pt.conn_prop_sessionTimeout), this.St.t(_.Ds, this.pt.conn_prop_mppLocal ? 1 : 0), 
        this.St.t(_.Ys, this.pt.conn_prop_rwSeparate === const_1.Const.RW_SEPARATE_DB_APPLY_WAIT ? const_1.Const.RW_SEPARATE_DB_APPLY_WAIT : 0 < this.pt.conn_prop_rwSeparate ? 1 : 0), 
        this.St.t(_.As, this.pt.newLobFlag ? 1 : 0), this.St.t(_.Ts, this.pt.conn_prop_osAuthType), 
        this.St.t(_.Zs, this.pt.longLobFlag ? 1 : 0), this.St.t(_.Cs, 1), this.br.lt;
        var i, s = bufferUtil_1.BufferUtil.o(this.pt.conn_prop_user, t), h = bufferUtil_1.BufferUtil.o(this.pt.conn_prop_pwd, t);
        if (s.length > const_1.Const.MAX_NAME_LEN) throw error_1.DBError.ECJS_USERNAME_TOO_LONG();
        if (h.length > const_1.Const.MAX_NAME_LEN) throw error_1.DBError.ECJS_PASSWORD_TOO_LONG();
        this.br.wt && this.pt.conn_prop_loginCertificate ? (s = (i = new rsaCipher_1.RSACipher(this.pt.conn_prop_loginCertificate)).encrypt(s), 
        h = i.encrypt(h)) : this.br.wt && (s = this.br.Ut.encrypt(s, !1), h = this.br.Ut.encrypt(h, !1)), 
        this.St.Xi(s), this.St.Xi(h), this.St.tn(this.pt.conn_prop_appName, t), this.St.tn(this.pt.conn_prop_osName, t), 
        this.St.tn(const_1.Const.HOST_NAME, t), this.St.zi(this.pt.rwStandby ? 1 : 0);
        var r = this.pt.loginOption.join(",");
        this.St.N(r, t);
    }, _.prototype.pn = function() {
        var t = this.pt.getServerEncoding();
        this.Ns(this.pt), this.St.Ei(t);
        var i, s, h, r, n, o = this.St.D();
        if (0 === o && 0 < this.pt.msgVersion ? this.pt.schema = this.pt.conn_prop_user.toUpperCase() : this.pt.schema = this.St.Di(o, t), 
        this.St.Ei(t), this.St.Ei(t), this.St.D(), this.St.D(), this.St.D(), this.pt.guid = this.St.Ei(t), 
        this.St.Ei(t), 1 === this.St.a(_.ks) && (this.pt.standbyHost = this.St.Ei(t), this.pt.standbyPort = this.St.D(), 
        this.pt.standbyCount = this.St.T()), 0 < this.St.ni(!1) && (this.pt.sessionID = this.St.gi()), 
        0 < this.St.ni(!1) && 1 == this.St.hi() && (this.pt.formatDate = "DD-MON-YY", this.pt.formatTime = "HH12.MI.SS.FF6 AM", 
        this.pt.formatTimestamp = "DD-MON-YY HH12.MI.SS.FF6 AM", this.pt.formatTimestampTZ = "DD-MON-YY HH12.MI.SS.FF6 AM +TZH:TZM", 
        this.pt.formatTimeTZ = "HH12.MI.SS.FF6 AM +TZH:TZM"), 0 < this.St.ni(!1) && ((i = this.St.Ii(t)) && (this.pt.formatDate = i), 
        (i = this.St.Ii(t)) && (this.pt.formatTime = i), (i = this.St.Ii(t)) && (this.pt.formatTimestamp = i), 
        (i = this.St.Ii(t)) && (this.pt.formatTimestampTZ = i), (i = this.St.Ii(t)) && (this.pt.formatTimeTZ = i)), 
        0 < this.St.ni(!1) && this.St.Ii(t), 0 < this.St.ni(!1) && this.St.ri(1, !1, !0), 
        this.pt.msgVersion >= msg_1.Msg.VERSION_ROWID_B12 && (s = 0 != this.St.vi(), h = this.St.vi(), 
        r = this.St.vi(), this.pt.rowidNBitsEpno = this.utl_rowid_get_n_bits_for_epno_org(h, s, r), 
        this.pt.rowidMaxHpno = this.utl_rowid_get_max_hpno_org(h, s, r), this.pt.rowidMaxEpno = this.utl_rowid_get_max_epno_org(this.pt.rowidNBitsEpno, s, r), 
        this.pt.rowidNBitsReal = this.utl_rowid_get_n_bits_for_real_org(h, s, r)), 0 < this.St.ni(!1) && this.St.ri(2, !1, !0), 
        0 < this.St.ni(!1)) {
            var e = this.St.bi();
            switch (Math.min(e, _.MAX_INI_COUNT)) {
              case 1:
                this.St.ri(4, !1, !0);
            }
        }
        return 0 < this.St.ni(!1) && this.pt.msgVersion >= msg_1.Msg.VERSION_NEW_SERVER_VERSION && (n = this.St.Ei(t), 
        _.processServerVersion(n, this.pt)), this.pt;
    }, _.prototype.Ns = function(t) {
        t.maxRowSize = this.St.s(_.zs), t.ddlAutoCommit = 1 == this.St.a(_.Es), t.isoLevel = connection_1.Connection.db2nodeIsoLevel(this.St.s(_.Gs)), 
        t.conn_prop_caseSensitive = 1 == this.St.a(_.Js), t.backSlashFlag = 1 == this.St.a(_.Ks), 
        t.svrStat = this.St.c(_.Ps), t.svrMode = this.St.c(_.Qs), t.constParaOpt = 1 == this.St.a(_.Rs), 
        t.dbTimezone = this.St.c(_.Vs), t.newLobFlag = 1 == this.St.a(_.Ws), t.longLobFlag = 1 == this.St.a(_.Xs), 
        this.St.a(_.$s), this.St.a(_.th), this.br.bt = 1e3 * this.St.H(_.ih), 0 === t.conn_prop_bufPrefetch && (t.conn_prop_bufPrefetch = this.St.s(_.sh)), 
        this.St.c(_.hh), t.dscControl = 1 === this.St.a(_.rh);
    }, _.prototype.utl_rowid_get_n_bits_for_epno_org = function(t, i, s) {
        var h = 1, r = 0;
        if (!i && 0 == s) return r;
        for (;h < t; ) h <<= 1, r++;
        return r;
    }, _.prototype.utl_rowid_get_max_hpno_org = function(t, i, s) {
        var h, r = i || 0 != s ? (h = this.utl_rowid_get_n_bits_for_epno_org(t, i, s), 0 != s ? (BigInt(1) << BigInt(20 - h)) - BigInt(1) : (BigInt(1) << BigInt(16 - h)) - BigInt(1)) : BigInt(65534);
        return r;
    }, _.prototype.utl_rowid_get_max_epno_org = function(t, i, s) {
        var h = i || 0 != s ? (1 << t) - 1 : 0;
        return h;
    }, _.prototype.utl_rowid_get_n_bits_for_real_org = function(t, i, s) {
        return 0 != s ? 44 : 48;
    }, _.Hs = msg_1.Msg.cs, _.qs = _.Hs + const_1.Const.ULINT_SIZE, _.Fs = _.qs + const_1.Const.ULINT_SIZE, 
    _.Bs = _.Fs + const_1.Const.ULINT_SIZE, _.Ss = _.Bs + const_1.Const.BYTE_SIZE, _.ws = _.Ss + const_1.Const.USINT_SIZE, 
    _.Ds = _.ws + const_1.Const.ULINT_SIZE, _.Ys = _.Ds + const_1.Const.BYTE_SIZE, _.As = _.Ys + const_1.Const.BYTE_SIZE, 
    _.Ts = _.As + const_1.Const.BYTE_SIZE, _.Zs = _.Ts + const_1.Const.BYTE_SIZE, _.Cs = _.Zs + const_1.Const.BYTE_SIZE, 
    _.nh = (_.zs = _.cs) + const_1.Const.ULINT_SIZE, _.Es = _.nh + const_1.Const.ULINT_SIZE, 
    _.Gs = _.Es + const_1.Const.BYTE_SIZE, _.Js = _.Gs + const_1.Const.ULINT_SIZE, _.Ks = _.Js + const_1.Const.BYTE_SIZE, 
    _.Qs = _.Ks + const_1.Const.BYTE_SIZE, _.Ps = _.Qs + const_1.Const.USINT_SIZE, _.Rs = _.Ps + const_1.Const.USINT_SIZE, 
    _.Vs = _.Rs + const_1.Const.BYTE_SIZE, _.ks = _.Vs + const_1.Const.USINT_SIZE, _.Ws = _.ks + const_1.Const.BYTE_SIZE, 
    _.sh = _.Ws + const_1.Const.BYTE_SIZE, _.hh = _.sh + const_1.Const.ULINT_SIZE, _.rh = _.hh + const_1.Const.USINT_SIZE, 
    _.Xs = _.rh + const_1.Const.BYTE_SIZE, _.$s = _.Xs + const_1.Const.BYTE_SIZE, _.ih = _.$s + const_1.Const.BYTE_SIZE, 
    _.th = 63, _.MAX_INI_COUNT = 1, _;
}(msg_1.Msg);

exports.Login = Login;
