// dmdb/src/net/msg/prepare.js
"use strict";

var __extends = this && this.__extends || function() {
    var r = function(t, e) {
        return (r = Object.setPrototypeOf || {
            __proto__: []
        } instanceof Array && function(t, e) {
            t.__proto__ = e;
        } || function(t, e) {
            for (var s in e) e.hasOwnProperty(s) && (t[s] = e[s]);
        })(t, e);
    };
    return function(t, e) {
        function s() {
            this.constructor = t;
        }
        r(t, e), t.prototype = null === e ? Object.create(e) : (s.prototype = e.prototype, 
        new s());
    };
}();

Object.defineProperty(exports, "__esModule", {
    value: !0
});

var execute_1 = require("./execute"), msg_1 = require("./msg"), const_1 = require("../../desc/const"), executeRetInfo_1 = require("../../desc/executeRetInfo"), parameter_1 = require("../../desc/parameter"), dbtype_1 = require("../../driver/dbtype"), typeDescriptor_1 = require("../../desc/typeDescriptor"), dm_1 = require("../../dm"), Prepare = function(o) {
    function t(t, e, s, r, i) {
        var n = o.call(this, t, msg_1.Msg.Hn, e, i) || this;
        return n.de = s, n.ye = r, n;
    }
    return __extends(t, o), t.prototype.jn = function() {
        var t, e = this.br.pt, s = this.dr, r = msg_1.Msg.cs;
        s.execOpts && (t = s.execOpts.autoCommit), void 0 === t && (t = dm_1.dmdb.autoCommit), 
        void 0 === t && (t = e.conn_prop_autoCommit), r += this.St.t(r, t ? 1 : 0), r += this.St.t(r, this.de ? 1 : 0), 
        r += this.St.t(r, 0), r += this.St.t(r, 1), r += this.St.t(r, e.compatibleOracle() ? 0 : 1), 
        r += this.St.B(r, this.ye), r += this.St.r(r, s.maxRows <= 0 || e.conn_prop_enRsCache ? const_1.Const.INT64_MAX : s.maxRows), 
        r += this.St.t(r, e.conn_prop_isBdtaRS ? const_1.Const.RS_BDTA : const_1.Const.RS_ROW), 
        r += this.St.B(r, 0), r += this.St.t(r, 0), r += this.St.t(r, 0), r += this.St.t(r, 0), 
        r += this.St.f(r, s.queryTimeout), this.St.t(r, s.innerExec ? 1 : 0), this.St.nn(s.nativeSql, e.getServerEncoding());
    }, t.prototype.pn = function() {
        if (this.de) return o.prototype.pn.call(this);
        var t = new executeRetInfo_1.ExecuteRetInfo(), e = this.br.pt, s = this.dr, r = msg_1.Msg.cs;
        t.retSqlType = this.St.c(r), r += const_1.Const.USINT_SIZE;
        var i = this.St.U(r);
        r += const_1.Const.USINT_SIZE;
        var n = this.St.c(r);
        if (r += const_1.Const.USINT_SIZE, this.St.i(r), r += const_1.Const.DDWORD_SIZE, 
        e.trxStatus = this.St.s(r), const_1.Const.ULINT_SIZE, 0 < i) {
            s.serverParams = this.be(i), s.bindParams = new Array(i);
            for (var h = 0; h < i; h++) {
                var _ = s.serverParams[h], c = new parameter_1.Parameter();
                c.type = _.type, c.prec = _.prec, c.scale = _.scale, c.nullable = _.nullable, c.hasDefault = _.hasDefault, 
                c.typeFlag = _.typeFlag, c.lob = _.lob, c.ioType = _.ioType, c.name = _.name, c.typeName = _.typeName, 
                c.tableName = _.tableName, c.schemaName = _.schemaName, c.lobTabId = _.lobTabId, 
                c.lobColId = _.lobColId, c.mask = _.mask, c.typeDescriptor = _.typeDescriptor, s.bindParams[h] = c;
            }
            s.paramCount = i;
        } else s.serverParams = new Array(0), s.bindParams = new Array(0), s.paramCount = 0;
        return s.columns = 0 < n ? this.hs(n, t.rsBdta) : new Array(0), t;
    }, t.prototype.be = function(t) {
        for (var e, s, r, i, n = new Array(t), h = 0; h < t; h++) {
            var _ = new parameter_1.Parameter();
            _.type = this.St.D(), _.prec = this.St.D(), _.scale = this.St.D(), _.nullable = 0 != this.St.D();
            var c = this.St.bi();
            _.hasDefault = 0 != (c & execute_1.Execute.vs), _.typeFlag = 0 != (c & execute_1.Execute.bs) ? parameter_1.Parameter.TYPE_FLAG_RECOMMEND : parameter_1.Parameter.TYPE_FLAG_EXACT, 
            _.lob = 0 != (c & execute_1.Execute.ds), this.St.D(), _.ioType = this.St.bi(), e = this.St.bi(), 
            s = this.St.bi(), r = this.St.bi(), i = this.St.bi(), _.name = this.St.Di(e, this.br.pt.getServerEncoding()), 
            _.typeName = this.St.Di(s, this.br.pt.getServerEncoding()), _.tableName = this.St.Di(r, this.br.pt.getServerEncoding()), 
            _.schemaName = this.St.Di(i, this.br.pt.getServerEncoding()), _.lob && (_.lobTabId = this.St.D(), 
            _.lobColId = this.St.bi()), _.type === dbtype_1.DBType.DATETIME || _.type === dbtype_1.DBType.DATETIME2 ? 0 != (_.scale & dbtype_1.DBType.LOCAL_DATETIME_SCALE_MASK) ? (_.scale = _.scale & ~dbtype_1.DBType.LOCAL_DATETIME_SCALE_MASK, 
            _.mask = const_1.Const.MASK_LOCAL_DATETIME) : 0 != (_.scale & dbtype_1.DBType.ORACLE_DATE_SCALE_MASK) && (_.scale = _.scale & ~dbtype_1.DBType.ORACLE_DATE_SCALE_MASK, 
            _.mask = const_1.Const.MASK_ORACLE_DATE) : _.type === dbtype_1.DBType.DECIMAL && _.scale === dbtype_1.DBType.ORACLE_FLOAT_SCALE_MASK ? (_.originalPrec = _.prec, 
            _.prec = Math.round(.30103 * _.prec), _.scale = -1, _.mask = const_1.Const.MASK_ORACLE_FLOAT) : _.type === dbtype_1.DBType.VARCHAR && _.prec === dbtype_1.DBType.BFILE_PREC && _.scale === dbtype_1.DBType.BFILE_SCALE && (_.mask = const_1.Const.MASK_BFILE), 
            n[h] = _;
        }
        for (var o, h = 0; h < t; h++) {
            dbtype_1.DBType.isComplexType(n[h].type, n[h].scale) && ((o = new typeDescriptor_1.TypeDescriptor(this.br.pt)).unpack(this.St), 
            n[h].typeDescriptor = o);
        }
        return n;
    }, t;
}(execute_1.Execute);

exports.Prepare = Prepare;
