// dmdb/src/net/msg/startup.js
"use strict";

var __extends = this && this.__extends || function() {
    var h = function(t, s) {
        return (h = Object.setPrototypeOf || {
            __proto__: []
        } instanceof Array && function(t, s) {
            t.__proto__ = s;
        } || function(t, s) {
            for (var i in s) s.hasOwnProperty(i) && (t[i] = s[i]);
        })(t, s);
    };
    return function(t, s) {
        function i() {
            this.constructor = t;
        }
        h(t, s), t.prototype = null === s ? Object.create(s) : (i.prototype = s.prototype, 
        new i());
    };
}();

Object.defineProperty(exports, "__esModule", {
    value: !0
});

var msg_1 = require("./msg"), const_1 = require("../../desc/const"), driver_1 = require("../../driver"), login_1 = require("./login"), Startup = function(h) {
    function n(t, s) {
        var i = h.call(this, t, msg_1.Msg._r, null, s) || this;
        return i.pt = t.pt, i;
    }
    return __extends(n, h), n.prototype.jn = function() {
        var t, s = this, i = this.pt.getServerEncoding();
        this._h = this.pt.conn_prop_uKeyName ? 1 : 0, this.St.f(n.ch, 0), this.St.f(n.fh, this.pt.conn_prop_compress), 
        this.St.t(n.gh, this.pt.conn_prop_loginEncrypt ? n.GEN_KEYPAIR_SIMPLE : 0), this.St.t(n.mh, this.pt.conn_prop_loginEncrypt ? n.COMM_ENC_SIMPLE : 0), 
        this.St.t(n.vh, this.pt.conn_prop_isBdtaRS ? const_1.Const.RS_BDTA : const_1.Const.RS_ROW), 
        this.St.t(n.lh, this.pt.conn_prop_compressID), this.St.t(n.dh, "" !== this.pt.conn_prop_loginCertificate ? 1 : 0), 
        this.St.t(n.bh, 0), this.St.t(n.ph, 1), this.St.e(n.qh, msg_1.Msg.VERSION), this.St.tn(driver_1.Driver.VERSION, i), 
        this.St.zi(0), this._h, this.pt.conn_prop_loginEncrypt && (t = this.br.ar(), this.St.Xi(t)), 
        "" !== this.pt.conn_prop_catalog && this.St.tn(this.pt.conn_prop_catalog, i), this.St.Ci(n.xh.length), 
        n.xh.forEach(function(t) {
            return s.St.Ci(t);
        });
    }, n.prototype.An = function() {
        this.St.Jt(msg_1.Msg.kt);
    }, n.prototype.pn = function() {
        switch (this.pt.sslEncrypt = this.St.s(n.jh), this.St.s(n.Oh), this.St.s(n.kh)) {
          case 1:
            this.pt.serverEncoding = const_1.Const.ENCODING_UTF8;
            break;

          case 2:
            this.pt.serverEncoding = const_1.Const.ENCODING_EUCKR;
            break;

          case 3:
            this.pt.serverEncoding = const_1.Const.ENCODING_GB18030;
        }
        this.pt.conn_prop_compress = this.St.s(n.wh);
        var t, s, i = this.St.a(n.Mh), h = this.St.a(n.Sh);
        return this.pt.conn_prop_isBdtaRS = 0 < this.St.a(n.yh), this.pt.conn_prop_compressID = this.St.a(n.Ah), 
        this.br.dt = 1 === this.St.a(n.zh), this.pt.newClientType = 1 == this.St.a(n.Bh), 
        this.pt.msgVersion = Math.min(msg_1.Msg.VERSION, this.St.U(n.MSG_RES_STARTUP_PROTOCOL_VERSION)), 
        this.Sn(), this.pt.msgVersion < msg_1.Msg.VERSION_NEW_SERVER_VERSION && (t = this.pt.getServerEncoding(), 
        s = this.St.Ei(t), login_1.Login.processServerVersion(s, this.pt)), this._h, this.Yt = -1, 
        this.br.gt = -1, this.br.wt = !1, this.br.mt = !1, 0 < h && (this.br.gt = this.St.D()), 
        i === n.GEN_KEYPAIR_SIMPLE && (-1 === this.br.gt ? this.br.wt = !0 : this.br.mt = !0, 
        this.Wt = this.St.xi()), h === n.COMM_ENC_SIMPLE && (this.Yt = this.St.D()), this.pt.msgVersion >= msg_1.Msg.VERSION_DETERMINE_ENCRYPT && (this.St.ri(8, !1, !0), 
        this.pt.algorighm = this.St.T()), this;
    }, n.ch = msg_1.Msg.cs, n.fh = n.ch + const_1.Const.ULINT_SIZE, n.gh = n.fh + const_1.Const.ULINT_SIZE, 
    n.mh = n.gh + const_1.Const.BYTE_SIZE, n.vh = n.mh + const_1.Const.BYTE_SIZE, n.lh = n.vh + const_1.Const.BYTE_SIZE, 
    n.dh = n.lh + const_1.Const.BYTE_SIZE, n.bh = n.dh + const_1.Const.BYTE_SIZE, n.ph = n.bh + const_1.Const.BYTE_SIZE, 
    n.qh = n.ph + const_1.Const.BYTE_SIZE, n.jh = msg_1.Msg.cs, n.Oh = n.jh + const_1.Const.ULINT_SIZE, 
    n.kh = n.Oh + const_1.Const.ULINT_SIZE, n.wh = n.kh + const_1.Const.ULINT_SIZE, 
    n.Ch = n.wh + const_1.Const.ULINT_SIZE, n.Mh = n.Ch + const_1.Const.ULINT_SIZE, 
    n.Sh = n.Mh + const_1.Const.BYTE_SIZE, n.yh = n.Sh + const_1.Const.BYTE_SIZE, n.Ah = n.yh + const_1.Const.BYTE_SIZE, 
    n.Dh = n.Ah + const_1.Const.BYTE_SIZE, n.zh = n.Dh + const_1.Const.BYTE_SIZE, n.Bh = n.zh + const_1.Const.USINT_SIZE, 
    n.MSG_RES_STARTUP_PROTOCOL_VERSION = n.Bh + const_1.Const.BYTE_SIZE, n.GEN_KEYPAIR_SIMPLE = 1, 
    n.GEN_KEYPAIR_ZB = 2, n.COMM_ENC_SIMPLE = 2, n.COMM_ENC_ZB = 3, n.xh = [ 1, 0 ], 
    n;
}(msg_1.Msg);

exports.Startup = Startup;
