// dmdb/src/net/msg/stmtFree.js
"use strict";

var __extends = this && this.__extends || function() {
    var r = function(t, n) {
        return (r = Object.setPrototypeOf || {
            __proto__: []
        } instanceof Array && function(t, n) {
            t.__proto__ = n;
        } || function(t, n) {
            for (var e in n) n.hasOwnProperty(e) && (t[e] = n[e]);
        })(t, n);
    };
    return function(t, n) {
        function e() {
            this.constructor = t;
        }
        r(t, n), t.prototype = null === n ? Object.create(n) : (e.prototype = n.prototype, 
        new e());
    };
}();

Object.defineProperty(exports, "__esModule", {
    value: !0
});

var msg_1 = require("./msg"), StmtFree = function(i) {
    function t(t, n, e) {
        var r = i.call(this, t, msg_1.Msg.Fn, null, e) || this;
        return r.me = n, r;
    }
    return __extends(t, i), t.prototype.jn = function() {}, t.prototype.En = function() {
        i.prototype.En.call(this), this.St.f(msg_1.Msg.Tn, this.me);
    }, t.prototype.pn = function() {}, t;
}(msg_1.Msg);

exports.StmtFree = StmtFree;
